### PR TITLE
refactor build.sh script.

### PR DESCRIPTION
**Pull request by suzp1984**
Refactor & Debug build.sh script:

1. Let platforms, ABIs, release/debug targets case insensitive:
```
./build.sh -p Android Debug
```
Above cmd will do nothing, because Android & Debug are case sensitive. So let those keywords case insensitive maybe a good idea.

2. Add error msg to remind the developer:
```
./build.sh -p android
./build.sh -p andr
./build.sh -p android bug
./build.sh -p android -q arm debug
```
Above error cmds will not has hints to remind the developer, so I add error msg to cover: missing targets, invalidate platform, invalidate target, invalidate ABI.

3. Let `ISSUE_DESKTOP_BUILD` as false

This var can be `false`, I just need to rearrange the build orders at the end of the script, and the only dependency is `run_tests`, I add build the desktop `Debug` target manually inside `run_tests`.

4. Debug Android samples build

```
./build.sh -p android -k sample-gltf-viewer debug
```
The above cmd will fail, the backtrace told me the cmd was not found in `out/release/filament/bin/?`, but I'm running the debug build, and the cause is missing property `com.google.android.filament.tools-dir`, so the groovy plugin `filament/android/buildSrc/src/main/groovy/FilamentPlugin.groovy` will use the default config one.
```
                def filamentToolsPath = providers
                        .gradleProperty("com.google.android.filament.tools-dir")
                        .forUseAtConfigurationTime().get()
```
I don't understand where it config the `forUseAtConfigurationTime()`, but the value is `out/release/filament/`.
So config `-Pcom.google.android.filament.tools-dir` will resolve this bug.

5. use `pushd` & `popd` to replace `cd`

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -539,6 +540,7 @@ function build_android {
             for sample in ${ANDROID_SAMPLES}; do
                 ./gradlew \
                     -Pcom.google.android.filament.dist-dir=../out/android-release/filament \
+                    -Pcom.google.android.filament.tools-dir=../../../out/release/filament \
```
Comment (romainguy) [inline build.sh:543]: This is unnecessary

File: build.sh [hunk 2]
```diff
@@ -930,30 +946,54 @@ if [[ "${ISSUE_CLEAN_AGGRESSIVE}" == "true" ]]; then
     build_clean_aggressive
 fi
 
-if [[ "${ISSUE_DESKTOP_BUILD}" == "true" ]]; then
```
Comment (romainguy) [inline build.sh:933]: Why did you reorder the tests?

File: build.sh [hunk 3]
```diff
@@ -817,6 +820,12 @@ while getopts ":hacCfgijmp:q:uvslwtedk:b" opt; do
                         ISSUE_DESKTOP_BUILD=true
                         ISSUE_WEBGL_BUILD=false
                     ;;
+                    *)
+                        echo "unknown platform ${platform}"
```
[the remark this answers is not in the document]
Comment (suzp1984) [reply build.sh]: done in this way:

```
            platforms=$(echo "${OPTARG}" | tr ',' '\n')
            for platform in ${platforms}
            do
                case $(echo "${platform}" | tr '[:upper:]' '[:lower:]') in
                     *)
                         echo "Unknown platform ${platform}"
```

File: build.sh [hunk 4]
```diff
@@ -817,6 +820,12 @@ while getopts ":hacCfgijmp:q:uvslwtedk:b" opt; do
...
+                        echo "make sure use the validate ones [desktop|android|ios|webgl|all]"
```
[the remark this answers is not in the document]
Comment (suzp1984) [reply build.sh]: done

File: build.sh [hunk 5]
```diff
@@ -910,6 +925,7 @@ fi
 
 shift $((OPTIND - 1))
 
+shopt -s nocasematch
```
[the remark this answers is not in the document]
Comment (suzp1984) [reply build.sh]: done in this way:

```
for arg; do
    if [[ $(echo "${arg}" | tr '[:upper:]' '[:lower:]') == "release" ]]; then
        ISSUE_RELEASE_BUILD=true
    elif [[ $(echo "${arg}" | tr '[:upper:]' '[:lower:]') == "debug" ]]; then
        ISSUE_DEBUG_BUILD=true
    else
        BUILD_CUSTOM_TARGETS="${BUILD_CUSTOM_TARGETS} ${arg}"
    fi
done
```

File: build.sh (hunk 2 above)
Comment (suzp1984) [reply build.sh:933]: I checked, and reordering the `builds` doesn't make sense. reverted.

File: build.sh (hunk 1 above)
Comment (romainguy) [reply build.sh:539]: Please resolve this

File: build.sh [hunk 6]
```diff
@@ -490,6 +490,7 @@ function build_android {
             for sample in ${ANDROID_SAMPLES}; do
                 ./gradlew \
                     -Pcom.google.android.filament.dist-dir=../out/android-debug/filament \
+                    -Pcom.google.android.filament.tools-dir=../../../out/debug/filament \
```
[the remark this answers is not in the document]
Comment (romainguy) [reply build.sh:489]: Please resolve this

**Review by romainguy — CHANGES_REQUESTED** [review]

**Comment by poweifeng** [comment]
I don't really understand the benefit of allowing case insensitivity. 

But I think giving an error when necessary flags like `debug` or `release` is not specified is a good idea. But probably we just need a more general approach to validating the arguments.

**Comment by suzp1984** [comment]
> I don't really understand the benefit of allowing case insensitivity.
> 
> But I think giving an error when necessary flags like `debug` or `release` is not specified is a good idea. But probably we just need a more general approach to validating the arguments.


Oh, there are duplicated codes to check the `debug` & `release` for platforms (desktop, android, ios, webgl). 
So, I add a function to reduce the duplication:

```
function check_debug_release_build {
    if [[ "${ISSUE_DEBUG_BUILD}" == "true" || "${ISSUE_RELEASE_BUILD}" == "true" ]]; then
        "$@";
    else
        echo "You must declare a debug or release target for $@ builds."
        echo ""
        exit 1
    fi
}
```

**Review by romainguy — APPROVED** [review]
Thank you!

**Pull request merged**